draco/analysis/sidereal.py

```diff
@@ -477,6 +479,201 @@ def _regrid(self, vis, weight, lsd):
+    ) -> containers.SiderealStreamRebin:
+        """Apply the gradient correction to the input dataset.
+
+        Parameters
+        ----------
+        sstream
+            sidereal day to apply a correction to
+
+        Returns
+        -------
+        sstream
+            Input sidereal day with gradient correction applied, if needed
+        """
+        # Allows a normal sidereal stream to pass through this task.
+        # Helpful for creating generic configs
+        try:
+            sera = sstream.datasets["effective_ra"][:].local_array
+        except KeyError:
+            self.log.info(
+                f"Dataset of type ({type(sstream)}) does not have an effective "
+                "ra dataset. No correction will be applied."
+            )
+            return sstream
+
+        # If the reference stream was never set, use the data stream
+        if self.sstream_ref is None:
+            self.sstream_ref = sstream
+
+        self.sstream_ref.redistribute("freq")
+        sstream.redistribute("freq")
+
+        try:
+            # If the reference dataset has an effective ra dataset, use this
+            # when calculating the gradient. This could be true if the reference
+            # and target datasets are the same
+            ref_ra = self.sstream_ref.datasets["effective_ra"][:].local_array
+        except KeyError:
+            # Use fixed ra, which should be regularly sampled
+            ref_ra = self.sstream_ref.ra
+
+        vis = sstream.vis[:].local_array
+        weight = sstream.weight[:].local_array
+
+        ref_vis = self.sstream_ref.vis[:].local_array
+        ref_weight = self.sstream_ref.weight[:].local_array
+
+        for fi in range(vis.shape[0]):
+            # Skip if entirely masked already
+            if not np.any(weight[fi]):
+                continue
+
+            # Depends on whether using effective or grid ra in reference
+            rra = ref_ra[fi] if ref_ra.ndim > 1 else ref_ra
+
+            for vi in range(vis.shape[1]):
+                # Skip if entire baseline is masked
+                if not np.any(weight[fi, vi]):
+                    continue
+
+                # Depends on whether the effective ra has baseline dependence
+                rra_i = rra[vi] if rra.ndim > 1 else rra
+
+                # Calculate the weighted gradient for each baseline, sampled
+                # at the reference RA points
+                norm = tools.invert_no_zero(ref_weight[fi, vi])
+                grad = norm * np.gradient(ref_vis[fi, vi] * ref_weight[fi, vi], rra_i)
+
+                # Apply the correction to estimate the sample value at the
+                # RA bin centre
+                vis[fi, vi] -= grad * (sera[fi, vi] - sstream.ra)
+
+        # Since the correction has been applied, remove the `effective_ra`
+        # dataset to save memory. Also, remove the reference stream object
+        # to avoid having up to three copies of the dataset in memory
+        del self.sstream_ref
```

Comment:
I was deleting it for the sake of memory, since I would end up making a copy of the sidereal stream. I think that your suggestion below to add the `effective_ra` dataset to the standard `SiderealStream` container would avoid having to do the copy, and would remove the need for this line